client/messaging/common-message-layout.test.tsx

```diff
@@ -0,0 +1,328 @@
+        >
+          : 
+        </message-layout__Separator>
+        <common-message-layout__Text>
+          hey
+           
+          <common-message-layout__UserMention>
+            <ConnectedUsername
+              isMention={true}
+              userId={123}
+            />
+          </common-message-layout__UserMention>
+          
+        </common-message-layout__Text>
+      </TimestampMessageLayout>
+    `)
+  })
+
+  test('message with a link before a mention', () => {
+    expect(doMatch('http://www.example.com go here <@123>')).toMatchInlineSnapshot(`
+      <TimestampMessageLayout
+        highlighted={false}
+        time={0}
+      >
+        <common-message-layout__Username>
+          <ConnectedUsername
+            userId={2}
+          />
+        </common-message-layout__Username>
+        <message-layout__Separator
+          aria-hidden={true}
+        >
+          : 
+        </message-layout__Separator>
+        <common-message-layout__Text>
+          <a
+            href="http://www.example.com"
+            rel="noopener nofollow"
+            target="_blank"
+          >
+            http://www.example.com
+          </a>
+           go here
+           
+          <common-message-layout__UserMention>
+            <ConnectedUsername
+              isMention={true}
+              userId={123}
+            />
+          </common-message-layout__UserMention>
+          
+        </common-message-layout__Text>
+      </TimestampMessageLayout>
+    `)
+  })
+
+  test('message with a link between mentions', () => {
+    expect(doMatch('hey <@123> see http://www.example.com go here <@123>')).toMatchInlineSnapshot(`
+      <TimestampMessageLayout
+        highlighted={false}
+        time={0}
+      >
+        <common-message-layout__Username>
+          <ConnectedUsername
+            userId={2}
+          />
+        </common-message-layout__Username>
+        <message-layout__Separator
+          aria-hidden={true}
+        >
+          : 
+        </message-layout__Separator>
+        <common-message-layout__Text>
+          hey
+           
+          <common-message-layout__UserMention>
+            <ConnectedUsername
+              isMention={true}
+              userId={123}
+            />
+          </common-message-layout__UserMention>
+           
+          see 
+          <a
+            href="http://www.example.com"
+            rel="noopener nofollow"
+            target="_blank"
+          >
+            http://www.example.com
+          </a>
+           go here
+           
+          <common-message-layout__UserMention>
+            <ConnectedUsername
+              isMention={true}
+              userId={123}
+            />
+          </common-message-layout__UserMention>
+          
+        </common-message-layout__Text>
+      </TimestampMessageLayout>
+    `)
+  })
+
+  test('message with a link containing a mention', () => {
+    expect(doMatch('http://www.<@123>.com')).toMatchInlineSnapshot(`
```

Comment:
I appreciate testing if there's a mention inside the link, but this is not a valid domain and thus this test's result actually seems incorrect to me. The test for this particular case should probably place the mention in the path of the URL (rather than the domain), and we should ensure that the link matching stuff ignores invalid characters.

Valid characters are: `[A-z0-9-]` (this may be better handled in the url matching tests, I dunno). Other restriction on that seems to be it must be at least 3 characters, the first character must be a letter or number, and the last character must be a letter or a number (so not a dash).

Reply:
Hmm, you're right that this is better handled in the link tests. I'll delete this test from here and open an issue to better handle link matching.